### DISCUSSION TITLE
db에 접근하는 로직이 service에 있는게 맞는 걸까요?

### DISCUSSION THREAD

**Discussion by anso33** [GhostPangPang/GhostPong#69] [General]
저의 아주 작고 소중한, 주워들은 지식들에 의거(ㅎㅎㅎㅎ) 디비에 요청을 하는 것은 repository, 혹은 repository를 구현한 ARepository에서 구현했습니다. 서비스에서는 repository에 구현한, 예를 들면 findByNickname 같은 메서드를 호출해서 사용했던 기억이 납니다. 
```
 ARepository.findByNickname(nickname);
```
그래서 서비스 단에서 쿼리문을 사용하는 것이 낯설게 느껴지는데요...
제 지식이 짧아서 그런건지 아님 배겐드 팀 여러분들도 낯설게 느껴지시는지 궁금해서 토론의 장을 열어봅니다.ㅎㅎ

@Kimhan-nah  @jis-kim

**Comment by anso33** [comment]
내일 클러스터에서 이야기 해보면 좋을 것 같아욤!

**Comment by Kimhan-nah** [comment]
service에서 repository 사용해서 접근하고 있는데 그럼 같은 상황 아닌가용??
nest 구조가 controller -> service이고
그래서 service에서 각 로직이 수행되는데
이때 디비 업데이트가 필요하니까 typeorm repository를 이용해서 디비 업데이트를 하는 걸로 알고있어용
그럼 말씀하신 부분 동일하게 처리되고 있는 거 아닌가용??

만약 서비스에서 처리하지 않으면 컨트롤러에서 하는 건가요?? 고거슨... 이상한디 ㅎ

**Reply by anso33** [reply]
엄... 제말은  뭔가 typeorm의 repository를 가져다가 Db에 날리는 쿼리문을 구현한 메서드만 따로 모아둔 Repository 클래스를 하나 더 두는 것은 어떤지 생각해본거였어요!! ㅋㅎㅋㅎ 뭔가 스프링이랑 병행하니까 자꾸 동일시하려고 하는 것 같긴 하네여 제가ㅋㅋㅋㅋㅋㅋ

**Reply by Kimhan-nah** [reply]
쿼리가 길어져서 또는 재사용하려고 하시는 건가용?? 어떤 이유로 분리하고 싶으신 거예용??
만약 길어져서 또는 재사용 때문이라면 typeorm에 query builder를 사용할 수 있는 걸로 알고 있숩니당!

그리고 만약 레포지토리 클래스를 하나 만들게 되면 그 클래스는 그냥 typeorm을 한번 더 감싼? 클래스밖에 안 되는데 잘 모르겠어욤 ㅎ

**Reply by Kimhan-nah** [reply]
흐음 저도 잘 몰라서 좀더 알아보겠습니당 헤헷

**Reply by Kimhan-nah** [reply]
아 말씀하신 부분이 typeorm custom repository랑 같은 것 같아요!
한번 확인해주세욤 ㅎ.ㅎ

**Comment by jis-kim** [comment]
지금 이미 레포지토리 패턴을 사용하는 중이라고 생각했습니다만... 🤔 쿼리빌더 사용 시엔 딱히 그런 것 같진 않네요. 그리고 저랑 소현님 둘 다 service 안에서 db접근로직은 함수로 분리해버려서 비슷한 느낌을 받는 듯 합니다.
만약 레포지토리 패턴을 사용한다면 새 클래스를 만들어서 `Repository<Entity>` 를 상속받는 식으로 가면 좋을 것 같습니다. (이거 쓰다가 한나님이 댓글을 다셨는데 언급주신 커스텀 레포지토리와 동일한 개념 같습니다 ㅎㅎ)
TypeORM 에서 제공하는 메소드는 그대로 쓰고, 추가 매핑이 필요하면 다른 메소드를 선언하는 식으로요! 근데 그러면 또... 어디까지가 비즈니스 로직이고 어디까지가 db 인지 ...😇..참....

**Comment by Kimhan-nah** [comment]
여러분들 우리 함수 분리 기준이 repository 사용과 유효성 체크 분리가 아니라,
단순하게 코드의 재사용성 + 가독성이라고 전제를 하고 말씀 드릴게여
결론부터 말씀드리면 저는 findUserById 라는 함수 안에서 userId에 대한 유효성 체크(checkExsitUser )를 해도 된다고 생각합니당!
물론 createUser에서는 해당 user가 없어야 생성하는 건데 이때는 다른 함수 사용하고 대부분의 경우 해당 id가 존재해야 valid하잖아요?? 저는 그런 경우가 압도적으로 많다고 생각해용
함수명이 애매하다면 findExsitUser 로 변경해도 나쁘지 않다고 생각해여 ㅎ.ㅎ
어차피 지금 findUserByID는 repository.findOneBy 함수만 호출하고 있어서 우리가 정리한 함수 분리 기준이라면 굳이 안 빼도 되잖아요?
‘재사용’을 위해서 분리한 거라면, 오히려 유효성 체크를 포함하는 게 맞다고 생각합니다!

라고 얘기를 나눴고 모두 동의하였습니다! 탕탕탕!
